### PR TITLE
gpio interrupt support

### PR DESCRIPTION
**Pull request by merbanan**
Working WIP code for gpio interrupts. I have tested this with the gpio-keys driver and it works fine.

The board I am using has the button on gpio 21.

What is missing is interrupt support for gpio 32-63. It is straight forward to expand the airoha_gpiochip structure with the needed registers and airoha_gpio_irq_handler() needs to check 2 status registers instead of only one.

### CODE REVIEW

File: target/linux/airoha/dts/en7581.dtsi [hunk 1]
```diff
@@ -227,6 +227,11 @@
 			      <0x0 0x1fbf0268 0x0 0x4>,
 			      <0x0 0x1fa2001c 0x0 0x50>,
 			      <0x0 0x1fa2018c 0x0 0x4>,
+			      <0x0 0x1fbf0208 0x0 0x4>, /* Interrupt status register */
+			      <0x0 0x1fbf0210 0x0 0x4>, /* Level 0 */
+			      <0x0 0x1fbf028C 0x0 0x4>, /* Level 1 */
+			      <0x0 0x1fbf020C 0x0 0x4>, /* Edge 0 */
+			      <0x0 0x1fbf0280 0x0 0x4>, /* Edge 1 */
```
Comment (LorenzoBianconi) [inline target/linux/airoha/dts/en7581.dtsi:237]: @merbanan IIUC we are missing irq support for bank1. Can you please add it as well and then we can see how to organize the code?

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c [hunk 2]
```diff
@@ -15,11 +17,16 @@
 #include <linux/pinctrl/pinconf-generic.h>
 #include <linux/pinctrl/pinmux.h>
 #include <linux/platform_device.h>
+#include <linux/of_irq.h>
```
Comment (LorenzoBianconi) [inline target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c]: @merbanan can you please put the include in alphabetic order?

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c [hunk 3]
```diff
@@ -2118,9 +2136,155 @@ static int airoha_pinctrl_add_gpiochip(struct airoha_pinctrl *pinctrl,
...
+static void airoha_gpio_irq_unmask(struct irq_data *data)
+{
+
+	struct airoha_gpiochip *ctrl = irq_data_get_irq_chip_data(data);
+	int pin = data->hwirq;
+	unsigned long flags;
+	u32 tmp;
+	u32 mask = BIT((pin % 16) * 2) | (BIT((pin % 16) * 2) << 1);
+
+	printk("airoha_gpio_irq_unmask: pin = %d mask = %x\n", pin, mask);
```
Comment (LorenzoBianconi) [inline target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c]: @merbanan can you please get rid of printk?

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c [hunk 4]
```diff
@@ -2118,9 +2136,155 @@ static int airoha_pinctrl_add_gpiochip(struct airoha_pinctrl *pinctrl,
...
+		ctrl->falling |= mask;
+		break;
+	case IRQ_TYPE_LEVEL_HIGH:
+		ctrl->level_high |= mask;
+		break;
+	case IRQ_TYPE_LEVEL_LOW:
+		ctrl->level_low |= mask;
+		break;
+	}
+	printk("ctrl->rising: %x\n", ctrl->rising);
```
Comment (LorenzoBianconi) [inline target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c]: @merbanan same here

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c [hunk 5]
```diff
@@ -2189,6 +2353,63 @@ static int airoha_pinctrl_probe(struct platform_device *pdev)
...
+			return PTR_ERR(pinctrl->gpiochip.level[1]);
+
+		pinctrl->gpiochip.edge[0] = devm_platform_ioremap_resource(pdev, index++);
+		if (IS_ERR(pinctrl->gpiochip.edge[0]))
+			return PTR_ERR(pinctrl->gpiochip.edge[0]);
+
+		pinctrl->gpiochip.edge[1] = devm_platform_ioremap_resource(pdev, index++);
+		if (IS_ERR(pinctrl->gpiochip.edge[1]))
+			return PTR_ERR(pinctrl->gpiochip.edge[1]);
+
```
Comment (LorenzoBianconi) [inline target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c:2398]: @merbanan I think we can have a better code architecture here. Let's add IRQ support for bank 1 and then we will evaluate

File: target/linux/airoha/dts/en7581.dtsi (hunk 1 above)
Comment (merbanan) [reply target/linux/airoha/dts/en7581.dtsi:237]: Added.

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c (hunk 3 above)
Comment (merbanan) [reply target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c]: Ack

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c (hunk 5 above)
Comment (merbanan) [reply target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c:2398]: Ack.

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c (hunk 4 above)
Comment (merbanan) [reply target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c]: Ack.

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c (hunk 2 above)
Comment (merbanan) [reply target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c]: Ack

**Comment by merbanan** [comment]
The changes for the second bank support is untested. It should work.

I have no idea what to do about the long list of registers, there will be even more after serdes muxing...

**Comment by LorenzoBianconi** [comment]
@merbanan: I applied pinctrl changes with some cosmetic reworks. Can you please test it again and rebase/add missing code?

**Pull request closed without merging**